### PR TITLE
PARQUET-708: account for "worst case scenario" in MaxBufferSize for bit_width > 1

### PR DESCRIPTION
**Pull request by xhochy**

### CODE REVIEW

File: src/parquet/util/rle-encoding.h [hunk 1]
```diff
@@ -171,8 +171,12 @@ class RleEncoder {
 
   /// Returns the maximum byte size it could take to encode 'num_values'.
   static int MaxBufferSize(int bit_width, int num_values) {
-    int bytes_per_run = BitUtil::Ceil(bit_width * MAX_VALUES_PER_LITERAL_RUN, 8.0);
```
Comment (majetideepak) [inline src/parquet/util/rle-encoding.h]: Should we add another variable `MIN_VALUES_PER_LITERAL_RUN` to 8 and use the same previous logic here ?

File: src/parquet/column/levels-test.cc [hunk 2]
```diff
@@ -206,4 +206,32 @@ TEST(TestLevelEncoder, MinimumBufferSize) {
...
+  // Test the worst case for bit_width=2 consisting of
+  // LiteralRun(size=8)
+  // RepeatedRun(size=8)
+  // LiteralRun(size=8)
+  // ...
+  const int kNumToEncode = 1024;
+
+  std::vector<int16_t> levels;
+  for (int i = 0; i < kNumToEncode; ++i) {
+    if ((i % 16) < 7) {
```
Comment (majetideepak) [inline src/parquet/column/levels-test.cc]: should  `< 8` or  ` <= 7` ?

File: src/parquet/column/levels-test.cc [hunk 3]
```diff
@@ -206,4 +206,32 @@ TEST(TestLevelEncoder, MinimumBufferSize) {
...
+      levels.push_back(0);
+    } else {
+      levels.push_back(1);
+    }
+  }
+
+  std::vector<uint8_t> output(
+      LevelEncoder::MaxBufferSize(Encoding::RLE, 2, kNumToEncode));
```
Comment (majetideepak) [inline src/parquet/column/levels-test.cc]: Add tests for other bit widths may be 1, 4, 8 ?

File: src/parquet/util/rle-encoding.h (hunk 1 above)
Comment (xhochy) [inline src/parquet/util/rle-encoding.h]: This is not the minimal number of values for a literal run but rather the threshold for checking if we do a literal or a repeated run.

File: src/parquet/column/levels-test.cc (hunk 2 above)
Comment (xhochy) [inline src/parquet/column/levels-test.cc]: No, this is `< 7` to force a literal run. This loop generates the sequence of literal and repeated runs as mentioned in the comment above.

File: src/parquet/column/levels-test.cc (hunk 3 above)
Comment (xhochy) [inline src/parquet/column/levels-test.cc]: Will do.

**Comment by majetideepak** [comment]
+1 LGTM

**Comment by wesm** [comment]
+1, thank you

**Pull request closed without merging**